### ISSUE TITLE
imageupdate automation rewrites a `items: []` manifest on a weird location

### ISSUE THREAD

**Issue by srgvg** [fluxcd/flux2#5064]
### Describe the bug

Incidently bumped into this; not sure or just a bug, or partlly a feature.
We got kustomization directory containing amongst others 1 yaml file:
```
apiVersion: v1
items:
- apiVersion: apps/v1
  kind: Deployment
    (...)
- apiVersion: v1
  kind: Service
    name: web-api
    (...)
kind: List
```
located at `clusters/cosmorak/applications/buut/web-api/webapi.yaml` in the git repo
Next to that an imageupdateautomation:
```
apiVersion: image.toolkit.fluxcd.io/v1beta2
kind: ImageUpdateAutomation
metadata:
  name: web-api-update-automation
  namespace: flux-system  # The namespace where Flux resources reside
spec:
  interval: 1m0s  # Check every 1 minute for image updates
  sourceRef:
    kind: GitRepository
    name: flux-system  # The name of the GitRepository resource (from kubectl output)
    namespace: flux-system  # The namespace of the GitRepository resource
  git:
    checkout:
      ref:
        branch: main  # The branch of your repository to track ops repo changes
    commit:
      author:
       (...)
     messageTemplate: 'update image to {{range .Updated.Images}}{{println .}}{{end}}'
    push:
      branch: main  # Push changes to the main branch of the ops repo
  update:
    path: ./clusters/cosmorak 
    strategy: Setters  # Automatically update the image tag in your manifest
```
What then happens, is the weird part: there's a tag to update in the a forementioned deployment, but instead of just updating it in the file  `clusters/cosmorak/applications/buut/web-api/webapi.yaml`, the controller creates two new files at another location:
- `clusters/cosmorak/buut/deployment_web-api.yaml`
- `clusters/cosmorak/buut/service_web-api.yaml`
with the respective contents - and the update image tag - for the deployment and the service.
`buut` seems to be the name of the namespace - I think/guess.

### Steps to reproduce

flux version `v2.4.0`

### Expected behavior

flux updates the original file `clusters/cosmorak/buut/service_web-api.yaml` and does not create new files on an unexpected location

### OS / Distro

Ubuntu 24.04

### Flux version

 flux version flux: v2.4.0 distribution: flux-v2.4.0 helm-controller: v1.1.0 image-automation-controller: v0.39.0 image-reflector-controller: v0.33.0 kustomize-controller: v1.4.0 notification-controller: v1.4.0 source-controller: v1.4.1

### Flux check

└➞  flux check
► checking prerequisites
✔ Kubernetes 1.30.5+k3s1 >=1.28.0-0
► checking version in cluster
✔ distribution: flux-v2.4.0
✔ bootstrapped: true
► checking controllers
✔ helm-controller: deployment ready
► ghcr.io/fluxcd/helm-controller:v1.1.0
✔ image-automation-controller: deployment ready
► ghcr.io/fluxcd/image-automation-controller:v0.39.0
✔ image-reflector-controller: deployment ready
► ghcr.io/fluxcd/image-reflector-controller:v0.33.0
✔ kustomize-controller: deployment ready
► ghcr.io/fluxcd/kustomize-controller:v1.4.0
✔ notification-controller: deployment ready
► ghcr.io/fluxcd/notification-controller:v1.4.0
✔ source-controller: deployment ready
► ghcr.io/fluxcd/source-controller:v1.4.1
► checking crds
✔ alerts.notification.toolkit.fluxcd.io/v1beta3
✔ buckets.source.toolkit.fluxcd.io/v1
✔ gitrepositories.source.toolkit.fluxcd.io/v1
✔ helmcharts.source.toolkit.fluxcd.io/v1
✔ helmreleases.helm.toolkit.fluxcd.io/v2
✔ helmrepositories.source.toolkit.fluxcd.io/v1
✔ imagepolicies.image.toolkit.fluxcd.io/v1beta2
✔ imagerepositories.image.toolkit.fluxcd.io/v1beta2
✔ imageupdateautomations.image.toolkit.fluxcd.io/v1beta2
✔ kustomizations.kustomize.toolkit.fluxcd.io/v1
✔ ocirepositories.source.toolkit.fluxcd.io/v1beta2
✔ providers.notification.toolkit.fluxcd.io/v1beta3
✔ receivers.notification.toolkit.fluxcd.io/v1
✔ all checks passed

### Git provider

github

### Container Registry provider

github

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by gecube** [comment]
I could reproduce this and traced it down to specific code, so posting the findings here. Confirmed on the reported flux v2.4.0 code path **and still present on v2.8.8** (image-automation-controller v1.1.4, e2e) and on image-automation-controller `v1.2.3` (unit-level repro with the controller's own `internal/update` code). A validated one-line fix is at the end.

## TL;DR

kyaml's `kio.ByteReader` silently **unwraps** a file whose single document is `kind: List` into its `items[]`. The unwrapped items bypass the code path that applies the reader's `config.kubernetes.io/path` annotation, so they come out with *no* path. `kio.LocalPackageWriter` then invents default paths for path-less nodes — `<namespace>/<lowercase-kind>_<name>.yaml` — which is exactly `clusters/cosmorak/buut/deployment_web-api.yaml` / `service_web-api.yaml` from the report. The original List file is never rewritten (its "file" key is the empty string), so it keeps the old tag forever.

## Mechanics (with code refs)

1. `ScreeningLocalReader` reads each screened file with `kio.ByteReader{SetAnnotations: {config.kubernetes.io/path: <rel-path>}}`:
   [internal/update/filereader.go#L119-L128 @v1.2.3](https://github.com/fluxcd/image-automation-controller/blob/v1.2.3/internal/update/filereader.go#L119-L128)
2. `ByteReader.Read` applies `SetAnnotations` inside `decode()` — but when the file's only document is `kind: List` (or `ResourceList`) with `items`, it discards the (annotated) wrapper and emits the raw items via `yaml.NewRNode(...)`, never running them through `decode()` → **no path/index annotations on the items**:
   [kyaml/kio/byteio_reader.go#L247-L268 @kyaml/v0.21.1](https://github.com/kubernetes-sigs/kustomize/blob/kyaml/v0.21.1/kyaml/kio/byteio_reader.go#L247-L268) (same logic in v0.17.x used by flux v2.4.0)
3. The setter filter groups nodes by their path annotation. All unwrapped items — from **every** List file under `update.path` — share the key `""`, so as soon as one item changes, *all* nodes with path `""` are selected for writing:
   [internal/update/setters.go#L216-L244 @v1.2.3](https://github.com/fluxcd/image-automation-controller/blob/v1.2.3/internal/update/setters.go#L216-L244)
4. `LocalPackageWriter.Write` starts with `kioutil.DefaultPathAndIndexAnnotation("", nodes)`, which assigns `path.Join(namespace, "<kind>_<name>.yaml")` to any node missing the annotation, and writes those files under `PackagePath`:
   [kyaml/kio/pkgio_writer.go#L40](https://github.com/kubernetes-sigs/kustomize/blob/kyaml/v0.21.1/kyaml/kio/pkgio_writer.go#L40), [kyaml/kio/kioutil/kioutil.go#L133-L136](https://github.com/kubernetes-sigs/kustomize/blob/kyaml/v0.21.1/kyaml/kio/kioutil/kioutil.go#L133-L136)

This reconstructs the report exactly: `update.path: ./clusters/cosmorak` + a single-doc List at `applications/buut/web-api/webapi.yaml` → the commit adds `clusters/cosmorak/buut/deployment_web-api.yaml` (with the new tag) and `clusters/cosmorak/buut/service_web-api.yaml`, and leaves the original untouched. The Service is written out too even though nothing in it changed — it merely shares the `""` path key with the updated Deployment.

## e2e repro on flux v2.8.8

Local cluster + gitea + plain OCI registry; repo seeded with the reporter's layout; ImagePolicy flips `1.0.0 → 1.0.1`; ImageUpdateAutomation with `update.path: ./clusters/cosmorak`, `strategy: Setters`. The automation pushes:

```
update image to 192.168.97.97:5000/webapi:1.0.1

 clusters/cosmorak/buut/deployment_web-api.yaml | 17 +++++++++++++++++
 clusters/cosmorak/buut/service_web-api.yaml    | 10 ++++++++++
 2 files changed, 27 insertions(+)
```

`clusters/cosmorak/applications/buut/web-api/webapi.yaml` still contains `image: ...:1.0.0 # {"$imagepolicy": ...}` after the commit.

## Follow-on damage (this is worse than "files in a weird location")

- **The Kustomization covering that path breaks immediately** after the automation commit, because the resources now exist twice (inside the List *and* as standalone files):
  ```
  kustomize build failed: accumulating resources: ... may not add resource with an
  already registered id: Deployment.v1.apps/web-api.buut
  ```
  i.e. GitOps delivery for the whole path stops until manual cleanup (verified on kustomize-controller v1.8.5).
- **Deleting the generated files doesn't help**: the List file still holds the old tag, so the next reconcile detects a change again and re-commits the same files. The only way out is converting the List file (or removing the marker).
- **Cross-contamination**: any in-scope update anywhere under `update.path` also rewrites the items of *every other* single-doc List file that passes screening — even ones whose markers are out of scope and whose values didn't change (verified in the unit test below: a second List referencing a different policy got exploded to `<ns>/deployment_<name>.yaml` with unchanged content).
- Commit message templates iterating `.Changed.FileChanges` see the change under an **empty file name** (`""`), since that's the key the change was recorded under.

## Unit-level repro (controller's own code, v1.2.3)

<details>
<summary>internal/update test reproducing the report and the cross-contamination</summary>

```go
package update

import (
	"os"
	"path/filepath"
	"testing"

	"github.com/go-logr/logr"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"

	reflectorv1 "github.com/fluxcd/image-reflector-controller/api/v1"
)

func TestListManifestRepro(t *testing.T) {
	tmp := t.TempDir()
	pkgDir := filepath.Join(tmp, "clusters", "cosmorak")
	appDir := filepath.Join(pkgDir, "applications", "buut", "web-api")
	if err := os.MkdirAll(appDir, 0o755); err != nil {
		t.Fatal(err)
	}
	manifest := `apiVersion: v1
kind: List
items:
- apiVersion: apps/v1
  kind: Deployment
  metadata:
    name: web-api
    namespace: buut
  spec:
    template:
      spec:
        containers:
        - name: app
          image: registry.local/webapi:1.0.0 # {"$imagepolicy": "flux-system:webapi"}
- apiVersion: v1
  kind: Service
  metadata:
    name: web-api
    namespace: buut
  spec:
    selector:
      app: web-api
`
	if err := os.WriteFile(filepath.Join(appDir, "webapi.yaml"), []byte(manifest), 0o644); err != nil {
		t.Fatal(err)
	}

	policy := reflectorv1.ImagePolicy{
		ObjectMeta: metav1.ObjectMeta{Name: "webapi", Namespace: "flux-system"},
		Status: reflectorv1.ImagePolicyStatus{
			LatestRef: &reflectorv1.ImageRef{Name: "registry.local/webapi", Tag: "1.0.1"},
		},
	}

	result, err := UpdateWithSetters(logr.Discard(), pkgDir, pkgDir, []reflectorv1.ImagePolicy{policy})
	if err != nil {
		t.Fatal(err)
	}
	t.Logf("result.FileChanges: %#v", result.FileChanges)

	filepath.Walk(tmp, func(p string, info os.FileInfo, err error) error {
		if !info.IsDir() {
			b, _ := os.ReadFile(p)
			rel, _ := filepath.Rel(tmp, p)
			t.Logf("=== %s ===\n%s", rel, b)
		}
		return nil
	})
}
```

Output on `v1.2.3` (unpatched):

```
result.FileChanges: map[""]…   // change recorded under the empty file name
=== clusters/cosmorak/applications/buut/web-api/webapi.yaml ===   // untouched, still 1.0.0
=== clusters/cosmorak/buut/deployment_web-api.yaml ===            // NEW, tag 1.0.1
=== clusters/cosmorak/buut/service_web-api.yaml ===               // NEW, unchanged content
```

The cross-contamination variant (two unrelated Lists, policy in scope for only one) additionally produces `ns-b/deployment_app-b.yaml` for the untouched second List.

</details>

## Fix (validated)

`ByteReader` has a switch for exactly this; the reader in the controller just doesn't set it:

```diff
 		rdr := &kio.ByteReader{
 			Reader:            bytes.NewBuffer(filebytes),
 			SetAnnotations:    annotations,
 			PreserveSeqIndent: true,
+			DisableUnwrapping: true,
 		}
```

(internal/update/filereader.go). With this, the List node keeps its path annotation, the setter still finds markers inside `items[]` (the setter walks the whole node tree), and:

- the file is updated **in place**, `kind: List` wrapper and formatting preserved;
- no cross-contamination (the unrelated List stays untouched);
- `result.FileChanges` is keyed by the real relative path.

`go test ./internal/update ./internal/policy` passes with the change. One behavioral note: the change identity in `.Changed.Objects` is then reported as the enclosing List object rather than the inner Deployment; propagating the file's path annotation onto unwrapped items instead would keep per-item identity, but then the writer rewrites the file as multi-doc YAML and silently drops the `List` wrapper — mutating files the user didn't ask to change. Keeping the document intact seems preferable.

Happy to send a PR with the one-liner + regression tests if maintainers agree with the approach.

Workaround meanwhile (as @makkes said): store the objects as a multi-doc file or separate files. Note the unwrap only triggers when the List is the file's *single* document — a List inside a multi-doc file is not affected.


cc @matheuscscp
